rust/processor/src/processors/marketplaces_processor.rs

```diff
@@ -0,0 +1,1048 @@
+            // Collection metaddatas parsed from events. The key is generated collection_id for token v1,
+            // and collection address for token v2.
+            let mut collection_metadatas: HashMap<String, CollectionMetadata> =
+                HashMap::new();
+
+            let mut collection_offer_filled_metadatas: HashMap<
+                String,
+                CollectionOfferEventMetadata,
+            > = HashMap::new();
+
+            // Loop through the events for the all activities
+            for (index, event) in inner.events.iter().enumerate() {
+                // get event type and contract address
+                let type_str_vec = event.type_str.as_str().split("::").collect::<Vec<&str>>();
+                let contract_addr = type_str_vec[0];
+                let event_type = type_str_vec[type_str_vec.len() - 1];
+
+                // Only the events in the set will be indexed
+                // TODO - We currently only index events from our own smart contracts
+                if !events_set.contains(event_type) || !contract_addr.eq(marketplace_contract_address) {
+                    continue;
+                }
+
+                let event_data = serde_json::from_str::<serde_json::Value>(&event.data).unwrap();
+                let activity_fee_schedule_id = event.key.as_ref().unwrap().account_address.as_str();
+                let price = event_data["price"]
+                    .as_str()
+                    .expect("Error: Price missing from marketplace activity")
+                    .parse::<BigDecimal>()
+                    .unwrap();
+                let activity_offer_or_listing_id =
+                    if let Some(listing) = event_data["listing"].as_str() {
+                        standardize_address(listing)
+                    } else if let Some(token_offer) = event_data["token_offer"].as_str() {
+                        standardize_address(token_offer)
+                    } else {
+                        // collection_offer
+                        standardize_address(event_data["collection_offer"].as_str().unwrap())
+                    };
+                
+                // Get the token and collection metadata from event.
+                let token_metadata = TokenMetadata::from_event(event, txn_version).unwrap();
+                if let Some(token_metadata) = token_metadata {
+                    token_metadatas.insert(token_metadata.get_token_address().unwrap(), token_metadata.clone());
+                }
+                let collection_metadata =
+                    CollectionMetadata::from_event(event, txn_version).unwrap();
+                if let Some(collection_metadata) = collection_metadata {
+                    collection_metadatas.insert(
+                        collection_metadata.get_collection_address().unwrap(),
+                        collection_metadata,
+                    );
+                }
```

Comment:
I realize the way the Python processor is written is not the "correct" way to write processors.. The Python processor tries to read the individual data fields from the event. It's less code but can be error-prone (eg. if the smart contract changes, it will break assumptions made in this indexer and we won't know about it). 

The right way would be to create a Rust representation of the events data model that matches exactly the format of the events we expect. The ANS processor is a good example of this. ([link](https://github.com/aptos-labs/aptos-indexer-processors/blob/a9e42dc9d80a69e9d02cd5f19c39e0f3f5cfbeb0/rust/processor/src/models/ans_models/ans_lookup_v2.rs#L273))
For each event type, we have a struct for it. It's more verbose but less error-prone.

Reply:
Yep, I have them on line 438 - 506. Each data models (activities, offer, listing, etc) has a `from_event` method which extra and build the corresponding model.

Though I could get the `token_metadata` and `collection_metadata` from line 438 - 506 as well, but I think having it separately (line 424 - 435) is more cleaner. Any thoughts?

Reply:
It's less error-prone to have `token_metadata` and `collection_metadata` be read from the `from_event`.

Same thing with `price` and `activity_offer_or_listing_id`